Build depends on internet/dns resolution and GCP credentials

<!-- Please keep the structure below, or your issue may be closed. -->

**Describe the bug**
The build currently requires internet and GCP credentials to be present in the env you are running it in.

**To Reproduce**
Steps to reproduce the behavior:
1. Run the build in an environment without internet or GCP credentials

**Expected behavior**
The build should pass

**Screenshots, Promtail config, or terminal output**
This test relies on GCP credentials being present
```
--- FAIL: TestNewGCPLogTarget (0.00s)
    --- FAIL: TestNewGCPLogTarget/defaults_to_pull_target (0.00s)
        target_test.go:114: 
                Error Trace:    target_test.go:114
                Error:          Received unexpected error:
                                pubsub(publisher): google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
                Test:           TestNewGCPLogTarget/defaults_to_pull_target
                Messages:       NewGCPLogTarget(&{0xc0009e5900 0xc0003e6a60 0xc0003e6a78 0xc0003e6a80 0xc0003e6a98 0xc0003e6aa0}, &{0xc0008476c8}, &{0xc0008557a0 [] {0 {0 0}} {0 0} {{} 4294967296 0} 0x238b5a0}, [], test_job, &{  {} false  {   0 0   0 0 {   } {   } false false false 0s 0s 0s 0s [] [] [] [] <nil> false 0 0 0 0s 0s 0s 0s 0s 0s false  { panic <nil>} <nil> false   false <nil> <nil> <nil> }})
```

These tests rely on internet connectivity (or at least DNS resolution)
```
--- FAIL: TestMemcachedJumpHashSelector_PickSever (0.00s)
    memcached_client_selector_test.go:41: 
        	Error Trace:	memcached_client_selector_test.go:41
        	Error:      	Received unexpected error:
        	            	lookup duckduckgo.com on [::1]:53: read udp [::1]:45074->[::1]:53: read: connection refused
        	Test:       	TestMemcachedJumpHashSelector_PickSever
FAIL
```

```
--- FAIL: TestMultiKVSetup (0.02s)
    --- FAIL: TestMultiKVSetup/distributor (0.00s)
        modules_test.go:156: 
            	Error Trace:	modules_test.go:156
            	Error:      	Received unexpected error:
            	            	No address found for [eth0 en0]
            	            	create distributor lifecycler
            	            	github.com/grafana/loki/pkg/distributor.New
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/distributor/distributor.go:149
            	            	github.com/grafana/loki/pkg/loki.(*Loki).initDistributor
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules.go:263
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:120
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestMultiKVSetup.func8
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:155
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	            	error initialising module: distributor
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:122
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestMultiKVSetup.func8
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:155
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	Test:       	TestMultiKVSetup/distributor
    --- FAIL: TestMultiKVSetup/ingester (0.00s)
        modules_test.go:156: 
            	Error Trace:	modules_test.go:156
            	Error:      	Received unexpected error:
            	            	No address found for [eth0 en0]
            	            	error initialising module: ingester
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:122
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestMultiKVSetup.func8
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:155
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	Test:       	TestMultiKVSetup/ingester
    --- FAIL: TestMultiKVSetup/all (0.00s)
        modules_test.go:156: 
            	Error Trace:	modules_test.go:156
            	Error:      	Received unexpected error:
            	            	No address found for [eth0 en0]
            	            	create distributor lifecycler
            	            	github.com/grafana/loki/pkg/distributor.New
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/distributor/distributor.go:149
            	            	github.com/grafana/loki/pkg/loki.(*Loki).initDistributor
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules.go:263
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:120
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestMultiKVSetup.func8
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:155
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	            	error initialising module: distributor
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:122
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestMultiKVSetup.func8
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:155
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	Test:       	TestMultiKVSetup/all
--- FAIL: TestIndexGatewayRingMode_when_TargetIsRead (0.01s)
    --- FAIL: TestIndexGatewayRingMode_when_TargetIsRead/IndexGateway_always_set_to_ring_mode_when_running_as_part_of_read_target (0.00s)
        modules_test.go:179: 
            	Error Trace:	modules_test.go:179
            	Error:      	Received unexpected error:
            	            	No address found for [eth0 en0]
            	            	invalid ring lifecycler config
            	            	github.com/grafana/loki/pkg/storage/stores/shipper/indexgateway.(*RingManager).startServerMode
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/storage/stores/shipper/indexgateway/ringmanager.go:114
            	            	github.com/grafana/loki/pkg/storage/stores/shipper/indexgateway.NewRingManager
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/storage/stores/shipper/indexgateway/ringmanager.go:99
            	            	github.com/grafana/loki/pkg/loki.(*Loki).initIndexGatewayRing
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules.go:1008
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:120
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestIndexGatewayRingMode_when_TargetIsRead.func1
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:172
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	            	new index gateway ring manager
            	            	github.com/grafana/loki/pkg/loki.(*Loki).initIndexGatewayRing
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules.go:1011
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:120
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestIndexGatewayRingMode_when_TargetIsRead.func1
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:172
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	            	error initialising module: index-gateway-ring
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:122
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestIndexGatewayRingMode_when_TargetIsRead.func1
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:172
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	Test:       	TestIndexGatewayRingMode_when_TargetIsRead/IndexGateway_always_set_to_ring_mode_when_running_as_part_of_read_target
--- FAIL: TestIndexGatewayClientConfig_when_TargetIsQuerierOrRead (0.00s)
    --- FAIL: TestIndexGatewayClientConfig_when_TargetIsQuerierOrRead/IndexGateway_client_is_endabled_when_running_read_target (0.00s)
        modules_test.go:259: 
            	Error Trace:	modules_test.go:259
            	Error:      	Received unexpected error:
            	            	No address found for [eth0 en0]
            	            	invalid ring lifecycler config
            	            	github.com/grafana/loki/pkg/storage/stores/shipper/indexgateway.(*RingManager).startServerMode
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/storage/stores/shipper/indexgateway/ringmanager.go:114
            	            	github.com/grafana/loki/pkg/storage/stores/shipper/indexgateway.NewRingManager
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/storage/stores/shipper/indexgateway/ringmanager.go:99
            	            	github.com/grafana/loki/pkg/loki.(*Loki).initIndexGatewayRing
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules.go:1008
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:120
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestIndexGatewayClientConfig_when_TargetIsQuerierOrRead.func2
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:252
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	            	new index gateway ring manager
            	            	github.com/grafana/loki/pkg/loki.(*Loki).initIndexGatewayRing
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules.go:1011
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:120
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestIndexGatewayClientConfig_when_TargetIsQuerierOrRead.func2
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:252
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	            	error initialising module: index-gateway-ring
            	            	github.com/grafana/dskit/modules.(*Manager).initModule
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:122
            	            	github.com/grafana/dskit/modules.(*Manager).InitModuleServices
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/vendor/github.com/grafana/dskit/modules/modules.go:92
            	            	github.com/grafana/loki/pkg/loki.TestIndexGatewayClientConfig_when_TargetIsQuerierOrRead.func2
            	            		/build/9qb8249llcpc8pa2yy82mzb8mf6i3qa8-source/pkg/loki/modules_test.go:252
            	            	testing.tRunner
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/testing/testing.go:1439
            	            	runtime.goexit
            	            		/nix/store/wy3nidvssv59q0yrxcmn8wyjk9xk6bxz-go-1.18.5/share/go/src/runtime/asm_amd64.s:1571
            	Test:       	TestIndexGatewayClientConfig_when_TargetIsQuerierOrRead/IndexGateway_client_is_endabled_when_running_read_target
```

--- a/clients/pkg/promtail/targets/gcplog/pull_target.go
+++ b/clients/pkg/promtail/targets/gcplog/pull_target.go
@@ -9,6 +9,7 @@ import (
 	"github.com/go-kit/log/level"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/model/relabel"
+	"google.golang.org/api/option"
 
 	"github.com/grafana/loki/clients/pkg/promtail/api"
 	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
@@ -49,10 +50,11 @@ func newPullTarget(
 	relabel []*relabel.Config,
 	jobName string,
 	config *scrapeconfig.GcplogTargetConfig,
+	clientOptions ...option.ClientOption,
 ) (*pullTarget, error) {
 	ctx, cancel := context.WithCancel(context.Background())
 
-	ps, err := pubsub.NewClient(ctx, config.ProjectID)
+	ps, err := pubsub.NewClient(ctx, config.ProjectID, clientOptions...)
 	if err != nil {
 		return nil, err
 	}

--- a/clients/pkg/promtail/targets/gcplog/target.go
+++ b/clients/pkg/promtail/targets/gcplog/target.go
@@ -5,6 +5,7 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/prometheus/prometheus/model/relabel"
+	"google.golang.org/api/option"
 
 	"github.com/grafana/loki/clients/pkg/promtail/api"
 	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
@@ -26,10 +27,11 @@ func NewGCPLogTarget(
 	relabel []*relabel.Config,
 	jobName string,
 	config *scrapeconfig.GcplogTargetConfig,
+	clientOptions ...option.ClientOption,
 ) (Target, error) {
 	switch config.SubscriptionType {
 	case "pull", "":
-		return newPullTarget(metrics, logger, handler, relabel, jobName, config)
+		return newPullTarget(metrics, logger, handler, relabel, jobName, config, clientOptions...)
 	case "push":
 		return newPushTarget(metrics, logger, handler, jobName, config, relabel)
 	default:

--- a/clients/pkg/promtail/targets/gcplog/target_test.go
+++ b/clients/pkg/promtail/targets/gcplog/target_test.go
@@ -10,6 +10,8 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/prometheus/model/relabel"
 	"github.com/stretchr/testify/assert"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
 
 	"github.com/grafana/loki/clients/pkg/promtail/api"
 	"github.com/grafana/loki/clients/pkg/promtail/client/fake"
@@ -105,7 +107,7 @@ func TestNewGCPLogTarget(t *testing.T) {
 			// Since the push target underlying http server registers metrics in the default registerer, we have to override it to prevent duplicate metrics errors.
 			prometheus.DefaultRegisterer = prometheus.NewRegistry()
 
-			got, err := NewGCPLogTarget(tt.args.metrics, tt.args.logger, tt.args.handler, tt.args.relabel, tt.args.jobName, tt.args.config)
+			got, err := NewGCPLogTarget(tt.args.metrics, tt.args.logger, tt.args.handler, tt.args.relabel, tt.args.jobName, tt.args.config, option.WithCredentials(&google.Credentials{}))
 			// If the target was started, stop it after test
 			if got != nil {
 				defer func() { _ = got.Stop() }()

--- a/go.mod
+++ b/go.mod
@@ -119,6 +119,7 @@ require (
 	github.com/prometheus/alertmanager v0.24.0
 	github.com/prometheus/common/sigv4 v0.1.0
 	github.com/willf/bloom v2.0.3+incompatible
+	golang.org/x/oauth2 v0.0.0-20220622183110-fd043fe589d2
 	golang.org/x/text v0.3.7
 )
 
@@ -265,7 +266,6 @@ require (
 	go4.org/intern v0.0.0-20211027215823-ae77deb06f29 // indirect
 	go4.org/unsafe/assume-no-moving-gc v0.0.0-20220617031537-928513b29760 // indirect
 	golang.org/x/mod v0.6.0-dev.0.20220106191415-9b9b3d81d5e3 // indirect
-	golang.org/x/oauth2 v0.0.0-20220622183110-fd043fe589d2 // indirect
 	golang.org/x/term v0.0.0-20210927222741-03fcf44c2211 // indirect
 	golang.org/x/tools v0.1.10 // indirect
 	golang.org/x/xerrors v0.0.0-20220609144429-65e65417b02f // indirect

--- a/pkg/loki/modules_test.go
+++ b/pkg/loki/modules_test.go
@@ -262,24 +262,14 @@ func TestIndexGatewayClientConfig_when_TargetIsQuerierOrRead(t *testing.T) {
 	})
 }
 
+const localhost = "localhost"
+
 func minimalWorkingConfig(t *testing.T, dir, target string) Config {
 	prepareGlobalMetricsRegistry(t)
 
 	cfg := Config{}
-	cfg.SchemaConfig = config.SchemaConfig{
-		Configs: []config.PeriodConfig{
-			{
-				IndexType:  config.StorageTypeInMemory,
-				ObjectType: config.StorageTypeFileSystem,
-				RowShards:  16,
-				Schema:     "v11",
-				From: config.DayTime{
-					Time: model.Now(),
-				},
-			},
-		},
-	}
 	flagext.DefaultValues(&cfg)
+
 	// Set to 0 to find any free port.
 	cfg.Server.HTTPListenPort = 0
 	cfg.Server.GRPCListenPort = 0
@@ -298,6 +288,29 @@ func minimalWorkingConfig(t *testing.T, dir, target string) Config {
 			},
 		},
 	}
+
+	cfg.SchemaConfig = config.SchemaConfig{
+		Configs: []config.PeriodConfig{
+			{
+				IndexType:  config.StorageTypeInMemory,
+				ObjectType: config.StorageTypeFileSystem,
+				RowShards:  16,
+				Schema:     "v11",
+				From: config.DayTime{
+					Time: model.Now(),
+				},
+			},
+		},
+	}
+
+	cfg.Common.InstanceAddr = localhost
+	cfg.Ingester.LifecyclerConfig.Addr = localhost
+	cfg.Distributor.DistributorRing.InstanceAddr = localhost
+	cfg.IndexGateway.Mode = indexgateway.SimpleMode
+	cfg.IndexGateway.Ring.InstanceAddr = localhost
+	cfg.CompactorConfig.CompactorRing.InstanceAddr = localhost
+
+	cfg.Ruler.Config.Ring.InstanceAddr = localhost
 	cfg.Ruler.Config.StoreConfig.Type = config.StorageTypeLocal
 	cfg.Ruler.Config.StoreConfig.Local.Directory = dir
 

--- a/pkg/storage/chunk/cache/memcached_client.go
+++ b/pkg/storage/chunk/cache/memcached_client.go
@@ -98,7 +98,7 @@ func (cfg *MemcachedClientConfig) RegisterFlagsWithPrefix(prefix, description st
 func NewMemcachedClient(cfg MemcachedClientConfig, name string, r prometheus.Registerer, logger log.Logger) MemcachedClient {
 	var selector serverSelector
 	if cfg.ConsistentHash {
-		selector = &MemcachedJumpHashSelector{}
+		selector = DefaultMemcachedJumpHashSelector()
 	} else {
 		selector = &memcache.ServerList{}
 	}

--- a/pkg/storage/chunk/cache/memcached_client_selector.go
+++ b/pkg/storage/chunk/cache/memcached_client_selector.go
@@ -20,8 +20,28 @@ import (
 // with consistent DNS names where the naturally sorted order
 // is predictable.
 type MemcachedJumpHashSelector struct {
-	mu    sync.RWMutex
-	addrs []net.Addr
+	mu              sync.RWMutex
+	addrs           []net.Addr
+	resolveUnixAddr UnixResolver
+	resolveTCPAddr  TCPResolver
+}
+
+type UnixResolver func(network, address string) (*net.UnixAddr, error)
+
+type TCPResolver func(network, address string) (*net.TCPAddr, error)
+
+func NewMemcachedJumpHashSelector(resolveUnixAddr UnixResolver, resolveTCPAddr TCPResolver) *MemcachedJumpHashSelector {
+	return &MemcachedJumpHashSelector{
+		resolveUnixAddr: resolveUnixAddr,
+		resolveTCPAddr:  resolveTCPAddr,
+	}
+}
+
+func DefaultMemcachedJumpHashSelector() *MemcachedJumpHashSelector {
+	return &MemcachedJumpHashSelector{
+		resolveUnixAddr: net.ResolveUnixAddr,
+		resolveTCPAddr:  net.ResolveTCPAddr,
+	}
 }
 
 // staticAddr caches the Network() and String() values from
@@ -63,13 +83,13 @@ func (s *MemcachedJumpHashSelector) SetServers(servers ...string) error {
 	naddrs := make([]net.Addr, len(sortedServers))
 	for i, server := range sortedServers {
 		if strings.Contains(server, "/") {
-			addr, err := net.ResolveUnixAddr("unix", server)
+			addr, err := s.resolveUnixAddr("unix", server)
 			if err != nil {
 				return err
 			}
 			naddrs[i] = newStaticAddr(addr)
 		} else {
-			tcpAddr, err := net.ResolveTCPAddr("tcp", server)
+			tcpAddr, err := s.resolveTCPAddr("tcp", server)
 			if err != nil {
 				return err
 			}

--- a/pkg/storage/chunk/cache/memcached_client_selector_test.go
+++ b/pkg/storage/chunk/cache/memcached_client_selector_test.go
@@ -2,6 +2,7 @@ package cache_test
 
 import (
 	"fmt"
+	"net"
 	"testing"
 
 	"github.com/bradfitz/gomemcache/memcache"
@@ -35,8 +36,32 @@ func TestNatSort(t *testing.T) {
 	require.Equal(t, expected, input)
 }
 
+var mockUnixResolver = func(network, address string) (*net.UnixAddr, error) {
+	return &net.UnixAddr{
+		Name: address,
+		Net:  network,
+	}, nil
+}
+
+var ips = map[string][]byte{
+	"google.com:80":     net.ParseIP("198.51.100.121"),
+	"duckduckgo.com:80": net.ParseIP("10.0.0.1"),
+	"microsoft.com:80":  net.ParseIP("172.12.34.56"),
+}
+
+var mockTCPResolver = func(network, address string) (*net.TCPAddr, error) {
+	return &net.TCPAddr{
+		IP:   ips[address],
+		Port: 0,
+		Zone: "",
+	}, nil
+}
+
 func TestMemcachedJumpHashSelector_PickSever(t *testing.T) {
-	s := cache.MemcachedJumpHashSelector{}
+	s := cache.NewMemcachedJumpHashSelector(
+		mockUnixResolver,
+		mockTCPResolver,
+	)
 	err := s.SetServers("google.com:80", "microsoft.com:80", "duckduckgo.com:80")
 	require.NoError(t, err)
 
@@ -60,7 +85,10 @@ func TestMemcachedJumpHashSelector_PickSever(t *testing.T) {
 }
 
 func TestMemcachedJumpHashSelector_PickSever_ErrNoServers(t *testing.T) {
-	s := cache.MemcachedJumpHashSelector{}
+	s := cache.NewMemcachedJumpHashSelector(
+		mockUnixResolver,
+		mockTCPResolver,
+	)
 	_, err := s.PickServer("foo")
 	require.Error(t, memcache.ErrNoServers, err)
 }

